rhino/src/main/java/org/mozilla/javascript/NativeObject.java

```diff
@@ -687,6 +689,37 @@ && isEnumerable(stringId, sourceObj)) {
                     return ScriptRuntime.wrapBoolean(ScriptRuntime.same(a1, a2));
                 }
 
+            case ConstructorId_groupBy:
+                {
+                    Object items = args.length < 1 ? Undefined.instance : args[0];
+                    Object callback = args.length < 2 ? Undefined.instance : args[1];
+
+                    Map<Object, List<Object>> groups =
+                            AbstractEcmaObjectOperations.groupBy(
+                                    cx,
+                                    scope,
+                                    f,
+                                    items,
+                                    callback,
+                                    AbstractEcmaObjectOperations.KEY_COERCION.PROPERTY);
+
+                    NativeObject obj = (NativeObject) cx.newObject(scope);
+                    obj.setPrototype(null);
+
+                    for (Map.Entry<Object, List<Object>> entry : groups.entrySet()) {
+                        Scriptable elements = cx.newArray(scope, entry.getValue().toArray());
+
+                        ScriptableObject desc = (ScriptableObject) cx.newObject(scope);
```

Comment:
One last nit -- you can (and should please) replace lines 712-717 with:

obj.defineProperty(entry.getKey(), elements, ScriptableObject.PERMANENT);

This is not only less code but it's more efficient, because we're not creating a new object and messing around in a hash table to figure out what attributes to set.

Reply:
The only issue with that is that the entry's key is an Object, which `defineProperty` doesn't allow for the first parameter.

Reply:
OK -- good point -- let's merge this!